### PR TITLE
refactor: customization file support

### PR DESCRIPTION
**Pull request by vitorhugods**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [X] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [X] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

The current code for supporting build-time customization of the application is hard to follow.

#### Context

Values that are defined in `default.json` are read and we generate `BuildConfig` constants based on it.

For example, `file_restriction_enabled` might be `false` in the `default.json` file, but some customer needs a custom build that has it set to `true`.
Based on the JSON files, and the `FeatureConfigs.kt` file, we will generate a `BuildConfig.FILE_RESTRICTION_ENABLED`, that can be used in runtime.

### Causes

- Half the code is in one place (`variants.gradle.kts`), the other half in the `Customization` file, and there's no clear boundary between them.
- The expected environment variables and how it works is a bit hard to follow, requiring a lot of source-code reading.
- It's annoying have to define some variables inside each flavor and some on the top-level, without a clear reason for what goes where.

For example:
```json
{
    "dev": {
        "mls_support_enabled": false
    },
    "staging": {
        "mls_support_enabled": true
    },
    "encrypt_proteus_storage": false
}
```

Why is MLS Support inside, and Encrypted Proteus Storage on the top level? Where in the code do I need to look to figure out what goes where?

### Solutions

- Refactor the code so it becomes easier to follow (naming, structure, smaller functions, examples in Kdocs and tests), making a `ConfigurationFileImporter` that has a single responsibility, and a `Customization.getBuildtimeConfiguration` function that bridges the importer with our project variables.
- Simplify the variables setup across flavors: variables can be set both at the top-level and overwritten inside each flavor if needed.

#### New Format

Flavors are defined inside the `flavors` object.
Flavors will use the values defined on the top-level by default.
The values can be overwritten inside each `flavor` object.

On this example below, all flavors will use `color:red` by default, as it's defined on the top-level. The flavor `cabbage` overwrites it with `green`.
```json
{
    "color": "red",
    "flavors": {
        "apple": { },
        "strawberry": { },
        "cabbage": {
            "color": "green"
        }
    }
}
```

#### What about `custom.json`

The format is the same, and it will overwrite whatever is defined on the `default.json`.

Assume the `default.json` is defined above, and the `custom.json` is here below:
```json
{
    "color": "orange",
    "flavors": {
        "apple": {
            "color": "blue"
        }
    }
}
```

Now all flavors have `color:orange`, except for `apple` which is `blue`.
The same logic applied to `default.json` repeats using `custom.json`.

#### Test Coverage

- [X] I have added automated test to this contribution

#### How to Test

Other than just using the `default.json`, the interaction with a customization JSON file can be seen by hard-coding a `customizationOption` when calling `val buildtimeConfiguration = getBuildtimeConfiguration(rootDir = rootDir, customizationOption = CustomizationOption.FromFile(file("someCustomFile.json")))` 

----
#### PR Post Merge Checklist for internal contributors

- [X] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-1652/2/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.1.0(6226107)(6226107)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.1.0(6226107)(6226107)-staging-compat.apk)

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-1652/2/) **failed**.

**Review by yamilmedina — APPROVED** [review]
Great job ! even tests for them. Impressive 🍰

**Review by MohamadJaara — APPROVED** [review]
great work 🚀, do we need to add the Server links to the default.json ?

**Comment by vitorhugods** [comment]
> great work 🚀, do we need to add the Server links to the default.json ?

Yes! I'm raising another PR that changes that Soon™ after this one.

**Review by trOnk12 — APPROVED** [review]
Awesome job 🥇

**Comment by AndroidBob** [comment]
Build [3](https://10.10.124.134/job/wire-android-reloaded/job/PR-1652/3/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.1.0(6233852)(6233852)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.1.0(6233852)(6233852)-staging-compat.apk)
